src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs

```diff
@@ -41,58 +45,106 @@ private static FileStreamStrategy ChooseStrategyCore(SafeFileHandle handle, File
             return EnableBufferingIfNeeded(strategy, bufferSize);
         }
 
-        private static FileStreamStrategy ChooseStrategyCore(string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options)
+        private static FileStreamStrategy ChooseStrategyCore(string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options, long allocationSize)
         {
             if (UseNet5CompatStrategy)
             {
-                return new Net5CompatFileStreamStrategy(path, mode, access, share, bufferSize, options);
+                return new Net5CompatFileStreamStrategy(path, mode, access, share, bufferSize, options, allocationSize);
             }
 
             WindowsFileStreamStrategy strategy = (options & FileOptions.Asynchronous) != 0
-                ? new AsyncWindowsFileStreamStrategy(path, mode, access, share, options)
-                : new SyncWindowsFileStreamStrategy(path, mode, access, share, options);
+                ? new AsyncWindowsFileStreamStrategy(path, mode, access, share, options, allocationSize)
+                : new SyncWindowsFileStreamStrategy(path, mode, access, share, options, allocationSize);
 
             return EnableBufferingIfNeeded(strategy, bufferSize);
         }
 
-        // TODO: we might want to consider strategy.IsPipe here and never enable buffering for async pipes
         internal static FileStreamStrategy EnableBufferingIfNeeded(WindowsFileStreamStrategy strategy, int bufferSize)
             => bufferSize == 1 ? strategy : new BufferedFileStreamStrategy(strategy, bufferSize);
 
-        internal static SafeFileHandle OpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options)
-            => CreateFileOpenHandle(path, mode, access, share, options);
+        internal static SafeFileHandle OpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long allocationSize)
+            => CreateFileOpenHandle(path, mode, access, share, options, allocationSize);
 
-        private static unsafe SafeFileHandle CreateFileOpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options)
+        private static unsafe SafeFileHandle CreateFileOpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long allocationSize)
         {
-            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = GetSecAttrs(share);
+            using (DisableMediaInsertionPrompt.Create())
+            {
+                Debug.Assert(path != null);
+
+                uint ntStatus = 0;
+                if (allocationSize > 0 && (access & FileAccess.Write) != 0 && mode != FileMode.Open && mode != FileMode.Append)
+                {
+                    string prefixedAbsolutePath = PathInternal.IsExtended(path) ? path : @"\??\" + Path.GetFullPath(path); // we might consider getting rid of this managed allocation
```

Comment:
We have a helper method in `PathInternal` to do this:

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/IO/PathInternal.Windows.cs#L82-L90

Note: `\\?\` and `\??\` are valid variants of device prefixes. There are great descriptions of both [here](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file), [here](https://stackoverflow.com/a/46019856/8816314) and also in the comment inside [`EnsureExtendedPrefix` method](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/IO/PathInternal.Windows.cs#L102-L125), called by the one shared above.

Reply:
For some reason, `NtCreateFile` works only with paths prefixed as `\??\`. When I switch to `EnsureExtendedPrefix` it adds `\?\` and `NtCreateFile` stops working.

Reply:
I have added a test and removed the allocation (cc @Jozkee)

Reply:
@adamsitnik interesting that only `\??\` works with `NtCreateFile`. It's strange, that P/Invoke's [documentation ](https://docs.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntcreatefile) does mention the prefix `\??\`, and links to [this ](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) document for additional info, but that document does not explain the specific purpose of `\??\`. I suspect it's for NT devices. I'll look more into this.